### ISSUE TITLE
posixlib utsname is hard broken & potentially dangerous to memory.

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#3276]
PR #3264 fixed some problems with posixlib utsname. It also woke up a ~~sleeping~~ , well otherwise occupied,
tardy reviewer.

I believe that the implementation of `utsname` is broken & dangerous and that this breakage long predates
PR #3264. Dangerous in the sense that it invites an operating system to write to j-random memory in the
process: i.e. outside the supplied buffer.

The issue is in the declarations:
```
@extern
object utsname {
  type _256 = Digit3[_2, _5, _6]
  private type str = CArray[Byte, _256]
  type utsname = CStruct5[str, str, str, str, str]
  @extern def uname(utsname: Ptr[utsname]): CInt = extern
}
```
The Open Group 2018 specification says two relevant things
```

The <sys/utsname.h> header shall define the structure utsname which shall include at least the following members:
<vertical ellipsis>
The character arrays are of unspecified size, but the data stored in them shall be terminated by a null byte.
```
[man 7 uname]
(https://man7.org/linux/man-pages/man2/uname.2.htmlhttps://man7.org/linux/man-pages/man2/uname.2.html)
has a long and relevant discussion about the size of the fields in current practice.

The "at least" portion of the Open Group Spec means that an operating system is free to declare and use fields
beyond the buffer supplied by the `utsobject` definition. I spent months/years chasing a similar problem in
`time.h`.  Much as I hate to say it, the solution is some "glue" code which checks that the buffer is large enough
on every supported OS.  (curlicues  and epicycles can be added to make it more theoretically pure).

###### Short term fix.
I believe the short term fix is to change the _256 to _257, as mentioned in `man 7 uname`, then add a 
comment describing where the magic _257 comes from and giving the `man 7 uname` URL.

The second comment in the same place would say that no known supported OS writes extra fields 
(do we know that?  I doubt it) but that some new OS might and spray memory.

(Later: Linux and possibly others declares and possibly write the POSIX-allowed domainname field.
That field is usually at the end of the struct utsname, so any system with a NIS or YP domain name
could potentially cause the memory spray feared above.)

One would then run a small C program in each of Linux, macOS, and any other currently supported
OS, to check that the size of the overall struct and each field declared in Scala matched or exceeded
the size in that OS.  This guards against current bogosity but not future change.

##### Long term fix.

The long term fix requires some study.  Are the fields in the C structure pointers to char (i.e. 8 or so bytes) or are they
actual allocated space?  I suspect the former.  

Then I think some "glue" code is needed.   I think, but do not know, that the Scala declaration will have to change.
(Better to be done in V5.0 than V5.1 or V6.0).

**Comment by LeeTibbert** [comment]
For reference: Linux declares the storage inline and also declares the POSIX-allowed GNU  `domainname`.
From a C program (using software up to date circa May, 2023)

```
sizeof(utsname): |390|
  - sizeof(utsname.sysname): |65|
  - sizeof(utsname.nodename): |65|
  - sizeof(utsname.release): |65|
  - sizeof(utsname.version): |65|
  - sizeof(utsname.machine): |65|
```
Also verified by inspection of  Linux 6.4-rep-RC1 source. 

This means that the glue code would need to check, at C compile time,
not only the total size, vs OS definition, but also the order and size of
each field.  Not hard (in theory), just grunt work needing doing.

I think the "Document in utsname.scala that the largest size (_257) is being
used and C glue code checks OS comparability" approach is the easiest
way to go here.  Over-allocates a few (Kilo) bytes, but probably saves a week of
devo time and downstream bugs. Someone needing a good term project
could probably come up with a more frugal approach.

The implication of the actual storage being allocated inline means that the
kernel would have done the copy into the storage. The presumptive C glue
code need not do that.

**Comment by LeeTibbert** [comment]
In thinking about this, I realize that it is even more broken than I thought. Every name
after the first is likely to be garbage.  

As a thought experiment,  say Linux is filling in 65 bytes for
the sysname field. It is probably skipping 7 bytes for alignment
and writing at (zero based) index 71.  That is well within
what Scala Native thinks is the sysname field.  So `nodename`
disappears.  Similar discussion for other fields. They are unlikely
to appear on their own, but may appear as garbage in other fields.

**Comment by LeeTibbert** [comment]
My brain hurts!  

There is a short term fix to much but not all of the bogosity here.  `uname.c` glue code 
currently exists in the SN repository, so change:
`@extern def uname(utsname: Ptr[utsname]): CInt = extern`  to
```
@name("scalanative_uname")
@extern def uname(utsname: Ptr[utsname]): CInt = extern
```

That should 
1) ensure that OS does not overwrite Scala Native memory
2) ensure that SN fields are filled in as expected.

It does not fix:
1) Magic number 256. What & why, in both utsname.scala & uname.c.
    Guarded by some _Static_assert statements in the existing `uname.c`, this
    could probably be reduced to 96 or even 80 bytes. Still too much for a 
    stackalloc, but more reasonable.

2) 
```
object uname {
  implicit class utsnameOps(
```
should probably be changed to something like, or better than:
```
object utsnameOps{
  implicit class utsnamePtrOps(
```

This would make the usual 2 include statements parallel & reduce cognitive burden:
``
import foo.utsname._
import foo.utsnameOps._
``

**Comment by LeeTibbert** [comment]
Noted in passing:  

macOS seems to return a longish (80?, 100?) character string for "version" field.
This OK with the remaining _256 field sizes but makes the case for .c guard code
if/when we try to reduce the SN allocated size.

Such development work should probably be done in parallel on macOS & linux.

**Comment by ekrich** [comment]
Note: that macOS defines the length as follows:
`#define _SYS_NAMELEN    256`

**Comment by LeeTibbert** [comment]
Eric, thank you for the data.  The magic number is probably macOS.   

The next time I get a chance to take a run at this, I will probably add some
_Static_assert code and be sure to test on macOS.

I'd really like to see the reasoning for the magic number in the .scala code.
Reverse engineering is a pain, especially for a method which was totally broken
to begin with.

I may also change to use `memccpy`. It was introduced in POSIX in 2004 and should
be in any OS variant SN supports.  The copies in the .c file are safe only if one _knows_
that the Scala Native fields are large enough. Right now, that happens by lore or,
if some takes your glibc path, happenstance.

Glad I no longer carry a support pager.

**Comment by LeeTibbert** [comment]
Eric,   I built on what you posted by incorporating the `_SYS_NAMELEN` information into the `utsname.scala`
and `uname.c` files.   I also added some guard code to the latter.   

I was not able to figure out, in the time give/stolen,  an Ops implicit to convert fields to Strings.  I did leave
a breadcrumbs in both the `utsname.scala` and `UtsnameTest.scala` files, showing the necessary `at(0)` idiom.
